flagd/pkg/service/flag-sync/sync-multiplexer.go

```diff
@@ -0,0 +1,193 @@
+	"github.com/open-feature/flagd/core/pkg/model"
+	"github.com/open-feature/flagd/core/pkg/store"
+)
+
+// Multiplexer abstract subscription handling and storage processing.
+// Flag configurations will be lazy loaded using reFill logic upon the calls to publish.
+type Multiplexer struct {
+	store   *store.Flags
+	sources []string
+
+	subs         map[interface{}]subscription            // subscriptions on all sources
+	selectorSubs map[string]map[interface{}]subscription // source specific subscriptions
+
+	allFlags      string            // pre-calculated all flags in store as a string
+	selectorFlags map[string]string // pre-calculated selector scoped flags in store as strings
+
+	mu sync.Mutex
+}
+
+type subscription struct {
+	id      interface{}
+	channel chan payload
+}
+
+type payload struct {
+	flags string
+}
+
+// NewMux creates a new sync multiplexer
+func NewMux(store *store.Flags, sources []string) (*Multiplexer, error) {
+	m := &Multiplexer{
+		store:         store,
+		sources:       sources,
+		subs:          map[interface{}]subscription{},
+		selectorSubs:  map[string]map[interface{}]subscription{},
+		selectorFlags: map[string]string{},
+	}
+
+	return m, m.reFill()
+}
+
+// Register a subscription
+func (r *Multiplexer) Register(id interface{}, source string, con chan payload) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if source != "" && !slices.Contains(r.sources, source) {
+		return fmt.Errorf("no flag watcher setup for source %s", source)
+	}
+
+	var initSync string
+
+	if source == "" {
+		// subscribe for flags from all source
+		r.subs[id] = subscription{
+			id:      id,
+			channel: con,
+		}
+
+		initSync = r.allFlags
+	} else {
+		// subscribe for specific source
+		s, ok := r.selectorSubs[source]
+		if ok {
+			s[id] = subscription{
+				id:      id,
+				channel: con,
+			}
+		} else {
+			r.selectorSubs[source] = map[interface{}]subscription{
+				id: {
+					id:      id,
+					channel: con,
+				},
+			}
+		}
+
+		initSync = r.selectorFlags[source]
+	}
+
+	// Initial sync
+	con <- payload{flags: initSync}
+	return nil
+}
+
+// Publish sync updates to subscriptions
+func (r *Multiplexer) Publish() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// perform a refill prior to publishing
+	err := r.reFill()
+	if err != nil {
+		return err
+	}
+
+	// push to all source subs
+	for _, sub := range r.subs {
+		sub.channel <- payload{r.allFlags}
+	}
+
+	// push to selector subs
+	for source, flags := range r.selectorFlags {
+		for _, s := range r.selectorSubs[source] {
+			s.channel <- payload{flags}
+		}
+	}
+
+	return nil
+}
+
+// Unregister a subscription
+func (r *Multiplexer) Unregister(id interface{}, selector string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var from map[interface{}]subscription
+
+	if selector == "" {
+		from = r.subs
+	} else {
+		from = r.selectorSubs[selector]
+	}
+
+	delete(from, id)
+}
+
+// GetALlFlags per specific source
+func (r *Multiplexer) GetALlFlags(source string) (string, error) {
```

Comment:
```suggestion
func (r *Multiplexer) GetAllFlags(source string) (string, error) {
```

Reply:
commit https://github.com/open-feature/flagd/pull/1237/commits/d7c5798048864dcc504fffafbdebbf61fbdf1938 fixes this